File: python/kimix_native/todo.py
"""kimix_native.todo — todo kernels: merge, status counts, plain-text summary.

Native implementations live in ``runtime_py.todo`` (compiled kernels, GIL
released). The pure-Python functions below mirror the reference algorithms
exactly so ``KIMIX_NATIVE_TODO=0`` yields identical output.
"""

from __future__ import annotations

import os
import sys
from typing import Any

from . import _native, use_native


def _canonical_status(v: Any) -> str:
    if not isinstance(v, str):
        raise ValueError(f"Invalid status '{v}'. Must be one of: pending, in_progress, done.")
    normalized = v.strip().lower().replace("-", "_")
    if normalized not in {"pending", "in_progress", "done"}:
        raise ValueError(f"Invalid status '{v}'. Must be one of: pending, in_progress, done.")
    return normalized


def _normalize_title(v: Any) -> str:
    if not isinstance(v, str):
        v = str(v)
    stripped = v.strip()
    if not stripped:
        raise ValueError("Title cannot be empty or contain only whitespace.")
    return stripped


def _normalize_notes(v: Any) -> str | None:
    if v is None:
        return None
    stripped = str(v).strip()
    return stripped if stripped else None


def _normalize_code(v: Any) -> str | None:
    return None if v is None else str(v)


def _todo_item(v: dict[str, Any]) -> dict[str, Any]:
    """Validate and canonicalize a raw todo dict."""
    return {
        "title": _normalize_title(v.get("title")),
        "status": _canonical_status(v.get("status")),
        "notes": _normalize_notes(v.get("notes")),
        "code": _normalize_code(v.get("code")),
    }


def _is_unfinished(item: dict[str, Any]) -> bool:
    return item["status"] != "done"


def _format_title_list(titles: list[str]) -> str:
    return "[" + ", ".join(f'"{t}"' for t in titles) + "]"


def _merge_one(old: dict[str, Any], new: dict[str, Any]) -> dict[str, Any]:
    """Produce an updated todo preserving old notes/code when new omits them."""
    return {
        "title": old["title"],
        "status": new["status"],
        "notes": new["notes"] if new["notes"] is not None else old["notes"],
        "code": new["code"] if new["code"] is not None else old["code"],
    }


def _merge_by_title_update(
    old_items: list[dict[str, Any]], new_items: list[dict[str, Any]]
) -> list[dict[str, Any]]:
    new_by_title = {t["title"]: t for t in new_items}
    merged: list[dict[str, Any]] = []
    seen: set[str] = set()

    for old in old_items:
        new = new_by_title.get(old["title"])
        if new is not None:
            merged.append(_merge_one(old, new))
        else:
            merged.append(old)
        seen.add(old["title"])

    for new in new_items:
        if new["title"] not in seen:
            merged.append(new)
            seen.add(new["title"])

    return merged


def _shell_kind() -> str:
    """Return the label used for !-prefixed code (matches the native default)."""
    return "pwsh" if sys.platform == "win32" else "bash"


def _format_kind(code: str) -> str:
    stripped = code.strip()
    if stripped.startswith("!"):
        return _shell_kind()
    lowered = stripped.lower()
    if lowered.endswith((".py", ".sh", ".ps1")) and os.path.isfile(stripped):
        return "file"
    return "inline"


def merge(
    old_items: list[dict[str, Any]],
    new_items: list[dict[str, Any]],
    mode: str,
    fuzzy_warnings: dict[str, list[str]] | None = None,
) -> dict[str, Any]:
    """Merge/update todo lists.

    Returns a dict with ``items``, ``warnings``, ``error``, ``regressed``,
    and ``archived`` keys. This mirrors the runtime_py.todo kernel exactly.
    """
    if fuzzy_warnings is None:
        fuzzy_warnings = {}

    if use_native("TODO") and _native is not None:
        return _native.todo.merge(old_items, new_items, mode, fuzzy_warnings)

    result: dict[str, Any] = {
        "items": [],
        "warnings": [],
        "error": None,
        "regressed": [],
        "archived": [],
    }

    # Canonicalize inputs (mirror the native binding's indexed error messages).
    old: list[dict[str, Any]] = []
    for i, x in enumerate(old_items):
        try:
            old.append(_todo_item(x))
        except ValueError as exc:
            result["error"] = f"Invalid todo at index {i}: {exc}"
            return result
    new = []
    for i, x in enumerate(new_items):
        try:
            new.append(_todo_item(x))
        except ValueError as exc:
            result["error"] = f"Invalid todo at index {i}: {exc}"
            return result

    # Mode validation.
    mode_norm = mode.strip().lower().replace("-", "_")
    is_append = mode_norm == "append"
    is_overwrite = mode_norm == "overwrite"
    is_force = mode_norm == "force_overwrite"
    if not (is_append or is_overwrite or is_force):
        result["error"] = (
            f"Invalid mode '{mode}'. Must be one of: append, overwrite, force_overwrite."
        )
        return result

    # Duplicate titles in incoming list.
    seen: dict[str, int] = {}
    duplicates: list[str] = []
    for item in new:
        seen[item["title"]] = seen.get(item["title"], 0) + 1
        if seen[item["title"]] == 2:
            duplicates.append(item["title"])
    if duplicates:
        duplicates.sort()
        result["error"] = f"Duplicate todo titles found: {_format_title_list(duplicates)}"
        return result

    # Branch on mode.
    if is_force:
        final = list(new)
    elif is_overwrite:
        if any(_is_unfinished(x) for x in old):
            unfinished = "\n".join(x["title"] for x in old if _is_unfinished(x))
            result["error"] = (
                "Error: Cannot overwrite todos while old todos are not all done. "
                "Use mode='force_overwrite' if you really want to discard unfinished work.\n"
                f"Unfinished:\n{unfinished}"
            )
            return result
        final = list(new)
    else:  # append
        if not new:
            if any(_is_unfinished(x) for x in old):
                unfinished = ", ".join(x["title"] for x in old if _is_unfinished(x))
                result["error"] = (
                    "Error: Cannot clear todos while old todos are not all done. "
                    f"Unfinished: {unfinished}\n"
                    "Next step: mark them done first, "
                    "or use mode='force_overwrite' to discard them intentionally."
                )
                return result
            final = []
        else:
            final = _merge_by_title_update(old, new)

    # Fuzzy warnings in new-item order.
    for item in new:
        result["warnings"].extend(fuzzy_warnings.get(item["title"], []))

    # Regression detection (unless force_overwrite).
    if not is_force and old:
        old_status_map = {x["title"]: x["status"] for x in old}
        clamped: list[dict[str, Any]] = []
        for item in final:
            if old_status_map.get(item["title"]) == "done" and item["status"] != "done":
                item = dict(item)
                item["status"] = "done"
                result["regressed"].append(item["title"])
            clamped.append(item)
        final = clamped

    # Archive completed todos that are no longer kept.
    kept_titles = {x["title"] for x in final}
    result["archived"] = [
        dict(x) for x in old if x["status"] == "done" and x["title"] not in kept_titles
    ]

    # Enforce single in_progress (auto-fix).
    if not is_force:
        fixed: list[dict[str, Any]] = []
        seen_in_progress = False
        for item in final:
            if item["status"] == "in_progress":
                if seen_in_progress:
                    item = dict(item)
                    item["status"] = "done"
                    result["warnings"].append(
                        f'Auto-fixed "{item["title"]}": set to done (only one item may be in_progress)'
                    )
                else:
                    seen_in_progress = True
            fixed.append(item)
        final = fixed

    result["items"] = final
    return result


def status_counts(items: list[dict[str, Any]]) -> dict[str, int]:
    """Count todos by canonical status."""
    if use_native("TODO") and _native is not None:
        return _native.todo.status_counts(items)

    counts: dict[str, int] = {"pending": 0, "in_progress": 0, "done": 0}
    for item in items:
        try:
            status = _canonical_status(item.get("status"))
        except ValueError:
            continue
        if status in counts:
            counts[status] += 1
    return counts


def format_summary(items: list[dict[str, Any]], max_items: int = 50) -> str:
    """Return the plain-text todo summary rendered by the reference tool."""
    if use_native("TODO") and _native is not None:
        return _native.todo.format_summary(items, max_items)

    try:
        todos = [_todo_item(x) for x in items]
    except ValueError:
        return ""

    selected = [t for t in todos if t["status"] in ("pending", "in_progress")]
    if not selected:
        return ""
    selected = selected[:max_items]

    display_status = {
        "pending": "pending",
        "in_progress": "in progress",
        "done": "done",
    }
    lines: list[str] = []
    for t in selected:
        line = f"- [{display_status[t['status']]}] {t['title']}"
        if t["code"]:
            line += f"  `[code: {_format_kind(t['code'])}]`"
        if t["status"] == "in_progress" and t["notes"]:
            line += f"  Notes: {t['notes']}"
        lines.append(line)
    return "\n".join(lines)

File: src/runtime/todo/todo.cpp
/*
 * todo.cpp — Todo-list merge, validation, status counts, and plain-text summary.
 */

#include <runtime/todo/todo.h>

#include <algorithm>
#include <cctype>
#include <filesystem>
#include <format>
#include <iterator>
#include <set>

namespace kimix {
namespace runtime {
namespace todo {

namespace {

void trim_inplace(string& s) noexcept {
    size_t start = 0;
    while (start < s.size() && std::isspace(static_cast<unsigned char>(s[start]))) {
        ++start;
    }
    if (start == s.size()) {
        s.clear();
        return;
    }
    size_t end = s.size();
    while (end > start && std::isspace(static_cast<unsigned char>(s[end - 1]))) {
        --end;
    }
    s.assign(s.begin() + static_cast<string::difference_type>(start),
             s.begin() + static_cast<string::difference_type>(end));
}

string lower_dash_to_underscore(string_view sv) noexcept {
    string out(sv.begin(), sv.end());
    trim_inplace(out);
    for (char& c : out) {
        c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
    }
    std::replace(out.begin(), out.end(), '-', '_');
    return out;
}

bool is_unfinished(const TodoItem& item) noexcept {
    return item.status != "done";
}

string format_title_list(const vector<string>& titles) {
    string out = "[";
    for (size_t i = 0; i < titles.size(); ++i) {
        if (i > 0) {
            out += ", ";
        }
        out += '"';
        out += titles[i];
        out += '"';
    }
    out += "]";
    return out;
}

template <typename... Args>
string kformat(std::format_string<Args...> fmt, Args&&... args) {
    auto text = std::format(fmt, std::forward<Args>(args)...);
    return string(text.begin(), text.end());
}

// Validate and canonicalize a list of items in place. Returns an error message
// on failure, or std::nullopt on success.
std::optional<string> validate_items(vector<TodoItem>& items) {
    for (auto& item : items) {
        trim_inplace(item.title);
        if (item.title.empty()) {
            return kimix::string("Title cannot be empty or contain only whitespace.");
        }
        auto canon = canonical_status(item.status);
        if (!canon) {
            return kformat("Invalid status '{}'. Must be one of: pending, in_progress, done.", item.status);
        }
        item.status = std::move(*canon);
    }
    return std::nullopt;
}

vector<TodoItem> merge_by_title_update(
    const vector<TodoItem>& old_items,
    const vector<TodoItem>& new_items) {
    vector<TodoItem> merged;
    merged.reserve(old_items.size() + new_items.size());

    unordered_map<string, const TodoItem*, string_hash> new_by_title;
    for (const auto& item : new_items) {
        new_by_title[item.title] = &item;
    }

    set<string> seen;
    for (const auto& old : old_items) {
        auto it = new_by_title.find(old.title);
        if (it != new_by_title.end()) {
            const TodoItem& new_item = *it->second;
            TodoItem updated;
            updated.title = old.title;
            updated.status = new_item.status;
            if (new_item.has_notes) {
                updated.has_notes = true;
                updated.notes = new_item.notes;
            } else {
                updated.has_notes = old.has_notes;
                updated.notes = old.notes;
            }
            if (new_item.has_code) {
                updated.has_code = true;
                updated.code = new_item.code;
            } else {
                updated.has_code = old.has_code;
                updated.code = old.code;
            }
            merged.push_back(std::move(updated));
        } else {
            merged.push_back(old);
        }
        seen.insert(old.title);
    }

    for (const auto& new_item : new_items) {
        if (seen.find(new_item.title) == seen.end()) {
            merged.push_back(new_item);
            seen.insert(new_item.title);
        }
    }

    return merged;
}

} // namespace

std::optional<string> canonical_status(string_view status) noexcept {
    string normalized = lower_dash_to_underscore(status);
    if (normalized == "pending" || normalized == "in_progress" || normalized == "done") {
        return normalized;
    }
    return std::nullopt;
}

MergeResult merge(
    const vector<TodoItem>& old_items,
    const vector<TodoItem>& new_items,
    string_view mode,
    const unordered_map<string, vector<string>, string_hash>& fuzzy_warnings) {
    MergeResult result;

    // ------------------------------------------------------------------
    // 1. Validate mode.
    // ------------------------------------------------------------------
    const string mode_norm = lower_dash_to_underscore(mode);
    const bool is_append = (mode_norm == "append");
    const bool is_overwrite = (mode_norm == "overwrite");
    const bool is_force = (mode_norm == "force_overwrite");
    if (!is_append && !is_overwrite && !is_force) {
        result.error = kformat("Invalid mode '{}'. Must be one of: append, overwrite, force_overwrite.", mode);
        return result;
    }

    // ------------------------------------------------------------------
    // 2. Validate and canonicalize input items.
    // ------------------------------------------------------------------
    vector<TodoItem> old_validated = old_items;
    vector<TodoItem> new_validated = new_items;
    if (auto err = validate_items(old_validated)) {
        result.error = std::move(*err);
        return result;
    }
    if (auto err = validate_items(new_validated)) {
        result.error = std::move(*err);
        return result;
    }

    // ------------------------------------------------------------------
    // 3. Duplicate titles in the incoming list.
    // ------------------------------------------------------------------
    {
        unordered_map<string, size_t, string_hash> seen;
        vector<string> duplicates;
        for (const auto& item : new_validated) {
            auto [it, inserted] = seen.try_emplace(item.title, 1);
            if (!inserted) {
                if (++it->second == 2) {
                    duplicates.push_back(item.title);
                }
            }
        }
        if (!duplicates.empty()) {
            std::sort(duplicates.begin(), duplicates.end());
            result.error = kformat("Duplicate todo titles found: {}", format_title_list(duplicates));
            return result;
        }
    }

    // ------------------------------------------------------------------
    // 4. Branch on write mode.
    // ------------------------------------------------------------------
    if (is_force) {
        result.items = new_validated;
    } else if (is_overwrite) {
        if (std::any_of(old_validated.begin(), old_validated.end(), is_unfinished)) {
            string unfinished;
            for (const auto& item : old_validated) {
                if (is_unfinished(item)) {
                    if (!unfinished.empty()) {
                        unfinished += '\n';
                    }
                    unfinished += item.title;
                }
            }
            string err;
            kimix::format_to(
                std::back_inserter(err),
                "Error: Cannot overwrite todos while old todos are not all done. "
                "Use mode='force_overwrite' if you really want to discard unfinished work.\n"
                "Unfinished:\n{}",
                unfinished);
            result.error = std::move(err);
            return result;
        }
        result.items = new_validated;
    } else { // append
        if (new_validated.empty()) {
            if (std::any_of(old_validated.begin(), old_validated.end(), is_unfinished)) {
                string unfinished;
                for (const auto& item : old_validated) {
                    if (is_unfinished(item)) {
                        if (!unfinished.empty()) {
                            unfinished += ", ";
                        }
                        unfinished += item.title;
                    }
                }
                string err;
                kimix::format_to(
                    std::back_inserter(err),
                    "Error: Cannot clear todos while old todos are not all done. "
                    "Unfinished: {}\n"
                    "Next step: mark them done first, "
                    "or use mode='force_overwrite' to discard them intentionally.",
                    unfinished);
                result.error = std::move(err);
                return result;
            }
            result.items.clear();
        } else {
            result.items = merge_by_title_update(old_validated, new_validated);
        }
    }

    // ------------------------------------------------------------------
    // 5. Append fuzzy warnings in new-item order.
    // ------------------------------------------------------------------
    for (const auto& item : new_validated) {
        auto it = fuzzy_warnings.find(item.title);
        if (it != fuzzy_warnings.end()) {
            result.warnings.insert(result.warnings.end(), it->second.begin(), it->second.end());
        }
    }

    // ------------------------------------------------------------------
    // 6. Regression detection (clamped back to done).
    // ------------------------------------------------------------------
    if (!is_force && !old_validated.empty()) {
        unordered_map<string, string, string_hash> old_status;
        for (const auto& item : old_validated) {
            // Keep the first occurrence if duplicates somehow exist.
            old_status.try_emplace(item.title, item.status);
        }

        for (auto& item : result.items) {
            auto it = old_status.find(item.title);
            if (it != old_status.end() && it->second == "done" && item.status != "done") {
                item.status = "done";
                result.regressed.push_back(item.title);
            }
        }
    }

    // ------------------------------------------------------------------
    // 7. Archive completed todos that are no longer kept.
    // ------------------------------------------------------------------
    {
        set<string> kept_titles;
        for (const auto& item : result.items) {
            kept_titles.insert(item.title);
        }
        for (const auto& item : old_validated) {
            if (item.status == "done" && kept_titles.find(item.title) == kept_titles.end()) {
                result.archived.push_back(item);
            }
        }
    }

    // ------------------------------------------------------------------
    // 8. Enforce a single in-progress item (auto-fix).
    // ------------------------------------------------------------------
    if (!is_force) {
        bool seen_in_progress = false;
        for (auto& item : result.items) {
            if (item.status == "in_progress") {
                if (seen_in_progress) {
                    item.status = "done";
                    result.warnings.push_back(
                        kformat("Auto-fixed \"{}\": set to done (only one item may be in_progress)",
                               item.title));
                } else {
                    seen_in_progress = true;
                }
            }
        }
    }

    return result;
}

unordered_map<string, size_t, string_hash> status_counts(const vector<TodoItem>& items) {
    unordered_map<string, size_t, string_hash> counts;
    counts["pending"] = 0;
    counts["in_progress"] = 0;
    counts["done"] = 0;
    for (const auto& item : items) {
        if (item.status == "pending" || item.status == "in_progress" || item.status == "done") {
            ++counts[item.status];
        }
    }
    return counts;
}

string format_summary(const vector<TodoItem>& items, size_t max_items) {
    vector<const TodoItem*> selected;
    selected.reserve(items.size());
    for (const auto& item : items) {
        if (item.status == "pending" || item.status == "in_progress") {
            selected.push_back(&item);
        }
    }

    if (selected.empty()) {
        return {};
    }
    if (selected.size() > max_items) {
        selected.resize(max_items);
    }

    string scratch;
    bool first = true;
    for (const auto* item : selected) {
        if (!first) {
            scratch += '\n';
        }
        first = false;

        string_view display_status = item->status;
        if (item->status == "in_progress") {
            display_status = "in progress";
        }

        kimix::format_to(std::back_inserter(scratch), "- [{}] {}", display_status, item->title);

        if (item->has_code && !item->code.empty()) {
            string stripped = item->code;
            trim_inplace(stripped);
            string kind_label;
            if (!stripped.empty() && stripped[0] == '!') {
#ifdef KIMIX_PLATFORM_WINDOWS
                kind_label = "pwsh";
#else
                kind_label = "bash";
#endif
            } else {
                string lower = stripped;
                for (char& c : lower) {
                    c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
                }
                bool is_file = false;
                try {
                    is_file = std::filesystem::is_regular_file(stripped);
                } catch (...) {
                    is_file = false;
                }
                if ((lower.ends_with(".py") || lower.ends_with(".sh") || lower.ends_with(".ps1")) && is_file) {
                    kind_label = "file";
                } else {
                    kind_label = "inline";
                }
            }
            kimix::format_to(std::back_inserter(scratch), "  `[code: {}]`", kind_label);
        }

        if (item->status == "in_progress" && item->has_notes && !item->notes.empty()) {
            kimix::format_to(std::back_inserter(scratch), "  Notes: {}", item->notes);
        }
    }

    return scratch;
}

} // namespace todo
} // namespace runtime
} // namespace kimix
